content/200-concepts/100-components/02-prisma-client/058-transactions.mdx

```diff
@@ -395,3 +395,71 @@ See the following resources:
+- The timing issues discussed in this section do not apply to CockroachDB and SQLite, because these databases only support the highest `Serializable` isolation level.
+
+</Admonition>
+
+When two or more transactions run concurrently in certain [isolation levels](https://www.prisma.io/dataguide/intro/database-glossary#isolation-levels), timing issues can cause conflicts, such as the violation of unique constraints. For example, consider the following sequence of events where Transaction A and Tranasaction B both attempt to execute a `deleteMany` and a `createMany` operation:
+
+1. Transaction A: `deleteMany` operation removes a set of rows.
+1. Transaction B: `deleteMany` operation removes a set of rows.
+1. Transaction B: `createMany` operation creates a new set of rows.
+1. Transaction B: The application commits transaction B.
+1. Transaction A: `createMany` operation. The newly-inserted rows conflict with the rows that transaction B added at step 4.
+
+This conflict can occur at the isolation level `ReadCommited`, which is the default isolation level in PostgreSQL and Microsoft SQL Server. To avoid this problem, you can set a higher isolation level (`RepeatableRead` or `Serializable`).
+
+From Prisma version 4.4.0, you do not need to set the isolation level in your database configuration. You can instead set the isolation level on a transaction. This overrides the isolation level on your database for that transaction. [Learn more about isolation levels in Prisma tranasctions](#transaction-isolation-level).
+
+To avoid transaction write conflicts and deadlocks:
+
+1. On your transaction, use the `isolationLevel` parameter to `Prisma.TransactionIsolationLevel.Serializable`.
+
+   This ensures that your application commits multiple concurrent or parallel transactions as if they were run serially. When a transaction fails due to a write conflict or deadlock, Prisma Client returns a [P2034 error](/reference/api-reference/error-reference#p2034).
+
+2. In your application code, add a retry around your transaction to handle any P2034 errors, as shown in this example:
+
+```ts
+import { Prisma, PrismaClient } from '@prisma/client'
+
+const prisma = new PrismaClient()
+async function main() {
+  const MAX_RETRIES = 5
+  let retries = 0
+
+  let result
+  while (retries < MAX_RETRIES) {
+    try {
+      result = await prisma.$transaction(
+        [
+          prisma.user.deleteMany({
+            where: {
+              /** args */
+            },
+          }),
+          prisma.post.createMany({
+            data: {
+              /** args */
+            },
+          }),
+        ],
+        {
+          isolationLevel: Prisma.TransactionIsolationLevel.Serializable,
+        }
+      )
+    } catch (error) {
+      if (error.code === 'P2034') {
+        retries++
+        continue
+      }
+      throw error
+    }
+  }
+}
+```
```

Comment:
(Out of scope for this PR)

This works, but is pretty impractical (as you would have to repeat this code for each and every transaction in your codebase.

This might be possible to do much nicer via a middleware instead that can add retry for _all_ transactions with a certain isolation level for example.